[MRG] ENH: Make normalization an explicit transformation.

dsquareindia opened the pull request:
Addresses issue #69. What this PR includes:
* Added support for ''l1' norm.
* Added norm parameter to Similarity constructor.
* Added norm parameter to matutils.unitvec.
* Created file models/normmodel.py to make normalization an explicit transformation.

TODO:
- [X] Check for input validation
- [X] Make test file test/test_norms.py
- [X] Add tests for input validation using assertRaises
- [X] Add documentation
- [X] Add more tests to testTransform to cover all input type validations

@piskvorky @tmylk Could you please tell me if I'm going on the right track?

dsquareindia commented:
@piskvorky @tmylk I've created `models/normmodel.py` in order to make normalization an explicit transformation. I hope I'm not going off on a tangent with this pr :P

dsquareindia commented:
I've added initial tests for just checking the logic. Input validation testing still remains.

tmylk commented on gensim/matutils.py:
```diff
@@ -332,9 +343,15 @@ def unitvec(vec):
         vec = vec.tocsr()
-        veclen = numpy.sqrt(numpy.sum(vec.data ** 2))
+        if norm == 'l0':
+            veclen = len(vec.data)
```
L_0 norm is not homogeneous so it is impossible to scale.

dsquareindia replied:
Yes true.... Could you please elaborate more on the scaling part? Will it be affecting efficiency since as the corpus size increases we cannot use the fact that `f(ax) = (a^k)f(x)`? Should I remove the support for zero "norm"?

tmylk replied:
Sorry to confuse you, meant scaling in mathematical not IT sense.

What number do we need to divide vector `a=(5,5)` by in order to have `l_0_norm(a)=1`?

dsquareindia replied:
Thanks for the explanation! I think it would be best to remove L0 norm then and keep the normalization just for l1 and l2. Should I proceed with it?

tmylk replied:
Yep. Thanks.

On Mon, Apr 4, 2016 at 8:04 PM, Devashish Deshpande <
notifications@github.com> wrote:

> In gensim/matutils.py
> <https://github.com/piskvorky/gensim/pull/649#discussion_r58429753>:
>
> >      if scipy.sparse.issparse(vec): # convert scipy.sparse to standard numpy array
> >          vec = vec.tocsr()
> > -        veclen = numpy.sqrt(numpy.sum(vec.data ** 2))
> > +        if norm == 'l0':
> > +            veclen = len(vec.data)
>
> Thanks for the explanation! I think it would be best to remove L0 norm
> then and keep the normalization just for l1 and l2. Should I proceed with
> it?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/piskvorky/gensim/pull/649/files/8b27eb4f46e0854fefdb4b55b6cea6c42c11a14d#r58429753>
>

tmylk commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,75 @@
+    """
+    def __init__(self, corpus=None, norm='l2'):
+        """
+        Compute the 'l0', 'l1' or 'l2' normalization by normalizing seperately
```
separately

tmylk commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,75 @@
+          norml2_{i, j} = (i / sqrt(sum(square(values in j))))
+        """
+        if norm not in ('l0', 'l1', 'l2'):
+            raise ValueError("%s is not a supported norm" % norm)
```
Code duplication.

dsquareindia commented:
@tmylk I've removed L0 norm and added some more tests to `test_normmodel`.

dsquareindia commented:
@tmylk @piskvorky I've added further tests to cover all input types. I think this pr is ready for review.

dsquareindia commented:
Removed merge conflicts.

tmylk commented:
@dsquareindia Hey, why did you close this pull request? It is a useful feature.

dsquareindia commented:
Sorry for that; I'll open it again.

dsquareindia commented:
Removed redundant use of corpus, rebased and squashed. Could someone please review this?

tmylk commented on gensim/matutils.py:
```diff
@@ -354,12 +372,14 @@ def unitvec(vec):
-            return list(vec)
+        if norm == 'l1':
+            length = float(sum(abs(val) for _, val in vec))
+            assert length > 0.0, "Document contains all zero entries"
```
What is the reason for this check?

dsquareindia replied:
Sorry I had just added that to make it symmetric with the L2 case. I should be returning the vector unchanged in case of 0 length right? Should I proceed with the change?

dsquareindia commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,69 @@
+        self.num_nnz = numnnz
+        self.norms = norms
+
+    def __getitem__(self, bow):
```
call it normalize. getitem a bit confusing.

tmylk commented on gensim/test/test_normmodel.py:
```diff
@@ -0,0 +1,134 @@
+        self.assertTrue(issparse(normalized))
+
+        # Check if output is correct
+        expected = numpy.array([[0.04761905, 0., 0.0952381],
```
Test description as doc string is best
h

dsquareindia commented on gensim/matutils.py:
```diff
@@ -354,12 +372,14 @@ def unitvec(vec):
+            return ret_normalized_vec(vec, length)
+        if norm == 'l2':
+            length = 1.0 * math.sqrt(sum(val ** 2 for _, val in vec))
+            assert length > 0.0, "sparse documents must not contain any explicit zero entries"
```
As done with other input types, I should be returning the vector unchanged in case of 0 length. Should I change this?

dsquareindia commented on gensim/matutils.py:
```diff
@@ -353,13 +371,14 @@ def unitvec(vec):
+            return ret_normalized_vec(vec, length)
+        if norm == 'l2':
+            length = 1.0 * math.sqrt(sum(val ** 2 for _, val in vec))
+            assert length > 0.0, "sparse documents must not contain any explicit zero entries"
```
Should I change this to returning the vector unchanged in case of zero length as we are doing with other input types?

dsquareindia commented:
@tmylk I have split the test suite into multiple tests to ease error checking. I have also removed the `__getitem__` transformation and have made it a function called `normalize()`. I hope now the pr is a bit better! :)

dsquareindia commented:
@piskvorky could you please review this? I think it's ready for merge.

piskvorky commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,69 @@
+        """
+        Calculates the norm by calling matutils.unitvec with the norm parameter.
+        """
+        logger.info("Normalizing...")
```
This log message is not very illuminating to users... try being more specific.

piskvorky commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,69 @@
+        numnnz, docno = 0, -1
+        for docno, bow in enumerate(corpus):
+            numnnz += len(bow)
+            norms[docno] = matutils.unitvec(bow, self.norm)
```
how about dropping `docno` and using `len(norms)` instead? Seems to simplify the logic a bit.

dsquareindia replied:
Sorry I didn't understand this. The norms dictionary itself is being constructed by iterating through all the docs in the corpus in this loop. Did you mean rather than using `enumerate(corpus)` I can simply iterate through the corpus and use `len(norms)` for indexing? While writing this, I personally felt that explicitly mentioning `docno` would enhance the readability of the code.....

tmylk replied:
We could just return a list, same as corpus.

piskvorky commented:
Only did a quick scan, looks nicely done, but I'll leave the thorough review to @tmylk.

My only nitpick would be `Normmodel` => `NormModel`, for readability. WDYT?

dsquareindia commented on gensim/models/normmodel.py:
```diff
@@ -0,0 +1,69 @@
+        self.num_nnz = numnnz
+        self.norms = norms
+
+    def normalize(self, bow):
```
@piskvorky should this rather be a `__getitem__()` since a transformation [here](https://radimrehurek.com/gensim/interfaces.html#gensim.interfaces.TransformationABC) is defined as something which acts like `transformation[doc]`.

piskvorky replied:
Sure. We can also let users use either -- a named function or `[ ]`.

tmylk commented on gensim/matutils.py:
```diff
@@ -319,30 +319,48 @@ def veclen(vec):
+    if norm not in ('l1', 'l2'):
+        raise ValueError("'%s' is not a supported norm" % norm)
+    else:
+        return norm
```
just return `(lambda x: unitvec(x,norm)`

tmylk commented on gensim/matutils.py:
```diff
@@ -319,30 +319,48 @@ def veclen(vec):
+
+def isnorm(norm):
+    if norm not in ('l1', 'l2'):
+        raise ValueError("'%s' is not a supported norm" % norm)
```
This check should be inside unitvec, at the end of if-else. Then will be no need for this `isnorm` function.

dsquareindia replied:
I had written it here to avoid code duplication in other modules which need to validate the input norm from user.

dsquareindia commented:
@tmylk @piskvorky I've addressed all the comments. Hopefully it's much better now! Could you please review :)

dsquareindia commented:
@tmylk @piskvorky thanks a ton for the reviews! :beers:

dsquareindia commented:
@piskvorky I was thinking of writing a notebook tutorial for this feature however I can't think of a practical use case for it. What would be the best way to go about this?

piskvorky commented:
Not sure... @tmylk ? Maybe indexing some corpus with/without normalization, showing how the returned "most similar" docs (or words, or embeddings... basically any vectors) differ?
